app/src/main/java/com/github/se/studybuddies/viewModels/SharedTimerViewModel.kt

```diff
@@ -3,98 +3,118 @@ package com.github.se.studybuddies.viewModels
 import android.os.CountDownTimer
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
-import com.github.se.studybuddies.database.DbRepository
-import com.github.se.studybuddies.database.ServiceLocator
+import com.github.se.studybuddies.data.TimerState
+import com.github.se.studybuddies.database.DatabaseConnection
+import kotlinx.coroutines.*
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
 import kotlinx.coroutines.launch
 
 class SharedTimerViewModel(
     private val groupUID: String,
-    private val db: DbRepository = ServiceLocator.provideDatabase()
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+    private val mainDispatcher: CoroutineDispatcher = Dispatchers.Main
 ) : ViewModel() {
+  private val db = DatabaseConnection()
   private val _timerValue = MutableStateFlow(0L)
   val timerValue: StateFlow<Long> = _timerValue
 
   private val _timerEnd = MutableStateFlow(false)
   val timerEnd: StateFlow<Boolean> = _timerEnd
 
-  private var isRunning = false
+  private val _running_local = MutableStateFlow(false)
+  val running_local: StateFlow<Boolean> = _running_local
+
+  private val _isRunning = MutableStateFlow(false)
+  val isRunning: StateFlow<Boolean> = _isRunning
+
   private var countDownTimer: CountDownTimer? = null
 
   init {
-    viewModelScope.launch { syncTimerWithFirebase() }
+
+    startListeningForTimerUpdates()
   }
 
-  private suspend fun syncTimerWithFirebase() {
-    val group = db.getGroup(groupUID)
-    val timerState = group.timerState
-    val remainingTime = timerState.endTime - System.currentTimeMillis()
-    if (remainingTime > 0) {
-      _timerValue.value = remainingTime
-      isRunning = timerState.isRunning
-      if (isRunning) {
-        setupTimer(remainingTime)
-      } else {
-        pauseTimer()
-      }
+  private fun startListeningForTimerUpdates() {
+    db.subscribeToGroupTimerUpdates(
+        groupUID, _timerValue, _isRunning, ioDispatcher, mainDispatcher, ::onTimerStateChanged)
+  }
+
+  private suspend fun onTimerStateChanged(timerState: TimerState) {
+
+    _timerValue.value = timerState.endTime
+
+    _isRunning.value = timerState.isRunning
+
+    if (_isRunning.value) {
+      startTimer()
     } else {
-      resetTimerValues()
+      pauseTimer()
+    }
+  }
+
+  fun setTimer(hours: Long = 0, minutes: Long = 0, seconds: Long = 0) {
+    val newTime = (hours * 3600 + minutes * 60 + seconds) * 1000
+    _timerValue.value = newTime
+
+    viewModelScope.launch { db.updateGroupTimer(groupUID, TimerState(newTime, _isRunning.value)) }
+  }
+
+  fun startTimer() {
+    // Timer is already running
+
```

Comment:
To resolve the bug when an user click multiples time on the start and it's block the app, you could simply add
`if (_running_local.value) return` to avoid the execution of the rest of the function